File: internal/hooks/gitstate_test.go
// Tests for gitstate.go — git state resolution, advisory composition, and
// Go/Node parity assertions.
// Uses package hooks (internal) to set gitRunner directly.
package hooks

import (
	"context"
	"errors"
	"strings"
	"testing"
)

// makeStubRunner builds a gitRunnerFn that maps a key substring to a return
// value or error. Keys that identify each probe:
//
//	"symbolic-ref"  → default-branch probe
//	"--show-current" → current-branch probe
//	"--porcelain"   → working-tree status probe
func makeStubRunnerGo(responses map[string]interface{}) gitRunnerFn {
	return func(ctx context.Context, args []string) (string, error) {
		for key, val := range responses {
			for _, arg := range args {
				if arg == key {
					switch v := val.(type) {
					case error:
						return "", v
					case string:
						return v, nil
					}
				}
			}
		}
		return "", errors.New("unexpected git args: " + strings.Join(args, " "))
	}
}

// ── resolveGitState unit tests ────────────────────────────────────────────────

// 2.1(a) default-branch probe errors → dirty probe still runs (non-nil).
func TestGitState_DefaultBranchFailsDirtyStillRuns(t *testing.T) {
	old := gitRunner
	defer func() { gitRunner = old }()

	gitRunner = makeStubRunnerGo(map[string]interface{}{
		"symbolic-ref":   errors.New("no remote HEAD"),
		"--show-current": "feat/my-feature",
		"--porcelain":    "M scripts/foo.js",
	})

	result := resolveGitState(context.Background())

	if result.DefaultBranch != nil {
		t.Errorf("defaultBranch: expected nil, got %q", *result.DefaultBranch)
	}
	if result.Dirty == nil {
		t.Error("dirty: expected non-nil (probe ran independently), got nil")
	}
	if result.Dirty != nil && !*result.Dirty {
		t.Errorf("dirty: expected true, got false")
	}
}

// 2.1(b) empty porcelain output → dirty = &false; status error → dirty = nil.
func TestGitState_EmptyPorcelainClean(t *testing.T) {
	old := gitRunner
	defer func() { gitRunner = old }()

	gitRunner = makeStubRunnerGo(map[string]interface{}{
		"symbolic-ref":   "origin/main",
		"--show-current": "main",
		"--porcelain":    "",
	})

	result := resolveGitState(context.Background())

	if result.Dirty == nil {
		t.Fatal("dirty: expected non-nil, got nil")
	}
	if *result.Dirty {
		t.Errorf("dirty: expected false, got true")
	}
}

func TestGitState_StatusProbeErrorDirtyNil(t *testing.T) {
	old := gitRunner
	defer func() { gitRunner = old }()

	gitRunner = makeStubRunnerGo(map[string]interface{}{
		"symbolic-ref":   "origin/main",
		"--show-current": "main",
		"--porcelain":    errors.New("git status failed"),
	})

	result := resolveGitState(context.Background())

	if result.Dirty != nil {
		t.Errorf("dirty: expected nil when probe fails, got %v", *result.Dirty)
	}
}

// 2.1(c) untracked-only line → dirty = &true.
func TestGitState_UntrackedFileDirtyTrue(t *testing.T) {
	old := gitRunner
	defer func() { gitRunner = old }()

	gitRunner = makeStubRunnerGo(map[string]interface{}{
		"symbolic-ref":   "origin/main",
		"--show-current": "main",
		"--porcelain":    "?? file.txt",
	})

	result := resolveGitState(context.Background())

	if result.Dirty == nil {
		t.Fatal("dirty: expected non-nil")
	}
	if !*result.Dirty {
		t.Errorf("dirty: expected true for untracked file, got false")
	}
}

// 2.1(d) context deadline exhausted → all fields nil.
func TestGitState_DeadlineExhausted(t *testing.T) {
	old := gitRunner
	defer func() { gitRunner = old }()

	gitRunner = func(ctx context.Context, args []string) (string, error) {
		return "", ctx.Err()
	}

	ctx, cancel := context.WithCancel(context.Background())
	cancel() // immediately cancelled

	result := resolveGitState(ctx)

	if result.DefaultBranch != nil {
		t.Errorf("defaultBranch: expected nil on deadline, got %q", *result.DefaultBranch)
	}
	if result.CurrentBranch != nil {
		t.Errorf("currentBranch: expected nil on deadline, got %q", *result.CurrentBranch)
	}
	if result.Dirty != nil {
		t.Errorf("dirty: expected nil on deadline, got %v", *result.Dirty)
	}
}

// current-branch probe fails → currentBranch nil, others unaffected.
func TestGitState_CurrentBranchProbeFailsOthersOk(t *testing.T) {
	old := gitRunner
	defer func() { gitRunner = old }()

	gitRunner = makeStubRunnerGo(map[string]interface{}{
		"symbolic-ref":   "origin/main",
		"--show-current": errors.New("detached HEAD"),
		"--porcelain":    "",
	})

	result := resolveGitState(context.Background())

	if result.CurrentBranch != nil {
		t.Errorf("currentBranch: expected nil, got %q", *result.CurrentBranch)
	}
	if result.DefaultBranch == nil || *result.DefaultBranch != "main" {
		t.Errorf("defaultBranch: expected 'main', got %v", result.DefaultBranch)
	}
	if result.Dirty == nil || *result.Dirty {
		t.Errorf("dirty: expected &false, got %v", result.Dirty)
	}
}

// Strips remote/ prefix from symbolic-ref output.
func TestGitState_StripRemotePrefix(t *testing.T) {
	old := gitRunner
	defer func() { gitRunner = old }()

	gitRunner = makeStubRunnerGo(map[string]interface{}{
		"symbolic-ref":   "origin/develop",
		"--show-current": "develop",
		"--porcelain":    "",
	})

	result := resolveGitState(context.Background())

	if result.DefaultBranch == nil || *result.DefaultBranch != "develop" {
		t.Errorf("defaultBranch: expected 'develop', got %v", result.DefaultBranch)
	}
}

// ── composeAdvisory tests ─────────────────────────────────────────────────────

func TestComposeAdvisory_DefaultOnly(t *testing.T) {
	falsy := false
	msg := composeAdvisory(true, &falsy, "main")
	if !strings.Contains(msg, "rama por defecto") {
		t.Errorf("default-only: expected 'rama por defecto' in %q", msg)
	}
	if !strings.Contains(msg, "main") {
		t.Errorf("default-only: expected branch name 'main' in %q", msg)
	}
	if strings.Contains(msg, "sin commitear") {
		t.Errorf("default-only: must NOT contain 'sin commitear' in %q", msg)
	}
}

func TestComposeAdvisory_DefaultOnlyNilDirty(t *testing.T) {
	// dirty=nil (probe failed) should not trigger dirty advisory.
	msg := composeAdvisory(true, nil, "main")
	if !strings.Contains(msg, "rama por defecto") {
		t.Errorf("expected 'rama por defecto' when dirty=nil: %q", msg)
	}
	if strings.Contains(msg, "sin commitear") {
		t.Errorf("must NOT contain 'sin commitear' when dirty=nil: %q", msg)
	}
}

func TestComposeAdvisory_DirtyOnly(t *testing.T) {
	truthy := true
	msg := composeAdvisory(false, &truthy, "feat/my-feature")
	if !strings.Contains(msg, "sin commitear") {
		t.Errorf("dirty-only: expected 'sin commitear' in %q", msg)
	}
	if strings.Contains(msg, "rama por defecto") {
		t.Errorf("dirty-only: must NOT contain 'rama por defecto' in %q", msg)
	}
}

func TestComposeAdvisory_Combined(t *testing.T) {
	truthy := true
	msg := composeAdvisory(true, &truthy, "main")
	if !strings.Contains(msg, "rama por defecto") {
		t.Errorf("combined: expected 'rama por defecto' in %q", msg)
	}
	if !strings.Contains(msg, "sin commitear") {
		t.Errorf("combined: expected 'sin commitear' in %q", msg)
	}
	if !strings.Contains(msg, "main") {
		t.Errorf("combined: expected branch name in %q", msg)
	}
}

// ── 2.1(e) Parity assertions ──────────────────────────────────────────────────
// For each canonical scenario, assert the expected permissionDecision matches
// what the Node runtime would produce (validated by Phase 3/5 Node tests).

// Scenario: default-branch only (clean feature branch for comparison).
func TestGitState_ParityScenario_DefaultBranchOnly(t *testing.T) {
	old := gitRunner
	defer func() { gitRunner = old }()

	gitRunner = makeStubRunnerGo(map[string]interface{}{
		"symbolic-ref":   "origin/main",
		"--show-current": "main",
		"--porcelain":    "",
	})

	result := resolveGitState(context.Background())
	onDefault := result.DefaultBranch != nil && result.CurrentBranch != nil &&
		*result.DefaultBranch == *result.CurrentBranch
	isDirty := result.Dirty != nil && *result.Dirty

	if !onDefault {
		t.Fatal("parity/default-only: onDefault should be true")
	}
	if isDirty {
		t.Fatal("parity/default-only: isDirty should be false")
	}
	// Expected permissionDecision: "ask" (mirroring Node 3.1(a))
	advisory := composeAdvisory(onDefault, result.Dirty, *result.CurrentBranch)
	if !strings.Contains(advisory, "rama por defecto") {
		t.Errorf("parity/default-only: advisory must contain 'rama por defecto': %q", advisory)
	}
}

// Scenario: dirty-only (feature branch).
func TestGitState_ParityScenario_DirtyOnly(t *testing.T) {
	old := gitRunner
	defer func() { gitRunner = old }()

	gitRunner = makeStubRunnerGo(map[string]interface{}{
		"symbolic-ref":   "origin/main",
		"--show-current": "feat/x",
		"--porcelain":    "M modified.js",
	})

	result := resolveGitState(context.Background())
	onDefault := result.DefaultBranch != nil && result.CurrentBranch != nil &&
		*result.DefaultBranch == *result.CurrentBranch
	isDirty := result.Dirty != nil && *result.Dirty

	if onDefault {
		t.Fatal("parity/dirty-only: onDefault should be false (feature branch)")
	}
	if !isDirty {
		t.Fatal("parity/dirty-only: isDirty should be true")
	}
	advisory := composeAdvisory(onDefault, result.Dirty, "")
	if !strings.Contains(advisory, "sin commitear") {
		t.Errorf("parity/dirty-only: advisory must contain 'sin commitear': %q", advisory)
	}
}

// Scenario: combined (default branch AND dirty).
func TestGitState_ParityScenario_Combined(t *testing.T) {
	old := gitRunner
	defer func() { gitRunner = old }()

	gitRunner = makeStubRunnerGo(map[string]interface{}{
		"symbolic-ref":   "origin/main",
		"--show-current": "main",
		"--porcelain":    "M modified.js",
	})

	result := resolveGitState(context.Background())
	onDefault := result.DefaultBranch != nil && result.CurrentBranch != nil &&
		*result.DefaultBranch == *result.CurrentBranch
	advisory := composeAdvisory(onDefault, result.Dirty, *result.CurrentBranch)

	if !strings.Contains(advisory, "rama por defecto") {
		t.Errorf("parity/combined: must contain 'rama por defecto': %q", advisory)
	}
	if !strings.Contains(advisory, "sin commitear") {
		t.Errorf("parity/combined: must contain 'sin commitear': %q", advisory)
	}
}

// ── Finding 3 remediation (RED) — hostile branch name sanitized in composeAdvisory ──

func TestComposeAdvisory_HostileBranchName_ControlCharsStripped(t *testing.T) {
	// Branch name with control characters including ESC (ANSI injection attempt)
	hostile := "main\x00\x1f\x1b[31mred\x1b[0m\r\ncontinued"
	msg := composeAdvisory(true, nil, hostile)

	// Control characters must not appear in the advisory
	for _, r := range msg {
		if (r >= 0x00 && r <= 0x1f) || r == 0x7f {
			t.Errorf("control char U+%04X found in advisory: %q", r, msg)
		}
	}
	// Advisory must still be non-empty
	if msg == "" {
		t.Error("advisory must be non-empty even for hostile branch name")
	}
}

func TestComposeAdvisory_LongBranchName_Truncated(t *testing.T) {
	longBranch := strings.Repeat("a", 200)
	msg := composeAdvisory(true, nil, longBranch)
	// 200-char branch name must not appear verbatim
	if strings.Contains(msg, longBranch) {
		t.Error("200-char branch name must be truncated")
	}
	// Must contain the truncated form (120 'a's) or an ellipsis
	if !strings.Contains(msg, strings.Repeat("a", 120)) && !strings.Contains(msg, "…") {
		t.Errorf("truncated advisory must contain 120 'a' chars or ellipsis: %q", msg)
	}
}

// Scenario: clean feature branch → no advisory fires.
func TestGitState_ParityScenario_CleanFeature(t *testing.T) {
	old := gitRunner
	defer func() { gitRunner = old }()

	gitRunner = makeStubRunnerGo(map[string]interface{}{
		"symbolic-ref":   "origin/main",
		"--show-current": "feat/x",
		"--porcelain":    "",
	})

	result := resolveGitState(context.Background())
	onDefault := result.DefaultBranch != nil && result.CurrentBranch != nil &&
		*result.DefaultBranch == *result.CurrentBranch
	isDirty := result.Dirty != nil && *result.Dirty

	if onDefault || isDirty {
		t.Errorf("parity/clean-feature: guard must NOT fire (onDefault=%v, isDirty=%v)", onDefault, isDirty)
	}
}

File: scripts/hooks/lib/git-state.test.js
"use strict";

const assert = require("node:assert/strict");
const test = require("node:test");

const { resolveGitState, isRiskyAction, composeAdvisory } = require("./git-state.js");

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Creates a stubbed git runner that maps known arg patterns to outputs or
 * Errors. The key is matched against the args array with `args.includes(key)`.
 *
 * Keys that distinguish the three git probes:
 *   "symbolic-ref"  → default-branch probe
 *   "--show-current" → current-branch probe
 *   "--porcelain"   → working-tree status probe
 */
function makeStubRunner(responses) {
  return (args) => {
    for (const [key, value] of Object.entries(responses)) {
      if (args.includes(key)) {
        if (value instanceof Error) throw value;
        return value;
      }
    }
    throw new Error(`Unexpected git args: ${args.join(" ")}`);
  };
}

// ---------------------------------------------------------------------------
// Task 1.1 — resolveGitState per-field fail-open
// ---------------------------------------------------------------------------

test("resolveGitState: default-branch probe fails – dirty probe still runs", () => {
  const runner = makeStubRunner({
    "symbolic-ref": new Error("no remote HEAD configured"),
    "--show-current": "feat/my-feature",
    "--porcelain": "M scripts/foo.js",
  });
  const result = resolveGitState(runner);
  assert.equal(result.defaultBranch, null, "defaultBranch should be null when probe fails");
  assert.equal(result.dirty, true, "dirty probe ran independently and got non-empty output");
});

test("resolveGitState: dirty probe fails – dirty field is null, not false", () => {
  const runner = makeStubRunner({
    "symbolic-ref": "origin/main",
    "--show-current": "main",
    "--porcelain": new Error("git status --porcelain failed"),
  });
  const result = resolveGitState(runner);
  assert.equal(result.dirty, null, "dirty should be null when status probe fails");
  assert.notStrictEqual(result.dirty, false, "dirty must not be false when probe fails");
});

test("resolveGitState: empty porcelain output – dirty is false", () => {
  const runner = makeStubRunner({
    "symbolic-ref": "origin/main",
    "--show-current": "main",
    "--porcelain": "",
  });
  const result = resolveGitState(runner);
  assert.strictEqual(result.dirty, false, "empty porcelain output means clean working tree");
});

test("resolveGitState: untracked-only output – dirty is true", () => {
  const runner = makeStubRunner({
    "symbolic-ref": "origin/main",
    "--show-current": "main",
    "--porcelain": "?? newfile.txt",
  });
  const result = resolveGitState(runner);
  assert.strictEqual(result.dirty, true, "untracked-only output triggers dirty:true");
});

test("resolveGitState: git binary absent – all fields null", () => {
  const runner = () => { throw new Error("git: command not found"); };
  const result = resolveGitState(runner);
  assert.equal(result.defaultBranch, null);
  assert.equal(result.currentBranch, null);
  assert.equal(result.dirty, null);
});

test("resolveGitState: current-branch probe fails – currentBranch null, others unaffected", () => {
  const runner = makeStubRunner({
    "symbolic-ref": "origin/main",
    "--show-current": new Error("detached HEAD"),
    "--porcelain": "",
  });
  const result = resolveGitState(runner);
  assert.equal(result.currentBranch, null, "currentBranch should be null on probe failure");
  assert.equal(result.defaultBranch, "main", "defaultBranch should still resolve");
  assert.strictEqual(result.dirty, false, "dirty probe ran independently");
});

test("resolveGitState: strips remote prefix from symbolic-ref output", () => {
  const runner = makeStubRunner({
    "symbolic-ref": "origin/develop",
    "--show-current": "develop",
    "--porcelain": "",
  });
  const result = resolveGitState(runner);
  assert.equal(result.defaultBranch, "develop", "origin/ prefix is stripped");
  assert.equal(result.currentBranch, "develop");
  assert.strictEqual(result.dirty, false);
});

test("resolveGitState: empty show-current (detached HEAD) → currentBranch null", () => {
  const runner = makeStubRunner({
    "symbolic-ref": "origin/main",
    "--show-current": "",
    "--porcelain": "",
  });
  const result = resolveGitState(runner);
  assert.equal(result.currentBranch, null, "empty show-current means detached HEAD");
});

// ---------------------------------------------------------------------------
// Task 1.2 — isRiskyAction
// ---------------------------------------------------------------------------

test("isRiskyAction: git commit command returns true", () => {
  assert.strictEqual(isRiskyAction(["git commit -m 'fix: test'"]), true, "git commit via -m");
  assert.strictEqual(isRiskyAction(["git commit --amend"]), true, "git commit --amend");
  assert.strictEqual(isRiskyAction(["git commit"]), true, "minimal git commit");
});

test("isRiskyAction: non-commit git commands return false", () => {
  assert.strictEqual(isRiskyAction(["git status"]), false);
  assert.strictEqual(isRiskyAction(["git log --oneline"]), false);
  assert.strictEqual(isRiskyAction(["git push"]), false);
});

test("isRiskyAction: empty or missing commands return false", () => {
  assert.strictEqual(isRiskyAction([]), false);
  assert.strictEqual(isRiskyAction(undefined), false);
});

test("isRiskyAction: file-write tools alone are no longer risky (no command payload)", () => {
  // The guard now behaves like a pre-commit check: file-write tools (Edit,
  // Write, etc.) never carry a command payload, so isRiskyAction([]) is the
  // only case that applies to them, and it must be false.
  assert.strictEqual(isRiskyAction([]), false);
});

// ---------------------------------------------------------------------------
// Task 1.3 — composeAdvisory message variants
// ---------------------------------------------------------------------------

test("composeAdvisory: default-only case contains branch name and 'rama por defecto'", () => {
  const result = composeAdvisory(true, false, "main");
  assert.ok(result.includes("main"), "should contain branch name 'main'");
  assert.ok(result.includes("rama por defecto"), "should contain 'rama por defecto'");
  assert.ok(!result.includes("sin commitear"), "should NOT contain 'sin commitear'");
});

test("composeAdvisory: default-only with null dirty also returns default advisory", () => {
  // When dirty probe failed (null), the guard still fires for default branch
  const result = composeAdvisory(true, null, "main");
  assert.ok(result.includes("rama por defecto"), "should contain 'rama por defecto'");
  assert.ok(!result.includes("sin commitear"), "dirty=null should not trigger dirty advisory");
});

test("composeAdvisory: dirty-only case contains 'sin commitear' and not 'rama por defecto'", () => {
  const result = composeAdvisory(false, true, "feat/my-feature");
  assert.ok(result.includes("sin commitear"), "should contain 'sin commitear'");
  assert.ok(!result.includes("rama por defecto"), "should NOT contain 'rama por defecto'");
});

test("composeAdvisory: combined case contains both 'rama por defecto' and 'sin commitear'", () => {
  const result = composeAdvisory(true, true, "main");
  assert.ok(result.includes("rama por defecto"), "combined must mention default branch");
  assert.ok(result.includes("sin commitear"), "combined must mention uncommitted changes");
  assert.ok(result.includes("main"), "combined must include branch name");
});

test("composeAdvisory: combined is a single string (not two separate strings)", () => {
  const result = composeAdvisory(true, true, "main");
  assert.equal(typeof result, "string", "result must be a single string");
  assert.ok(result.length > 0, "result must be non-empty");
});

// ---------------------------------------------------------------------------
// Finding 1 remediation (RED) — shared 5s deadline across all three probes
// ---------------------------------------------------------------------------

test("resolveGitState: shared deadline — each probe receives a positive numeric timeoutMs as 2nd arg", () => {
  const capturedBudgets = [];
  const spyRunner = (args, timeoutMs) => {
    capturedBudgets.push(timeoutMs);
    if (args.includes("symbolic-ref")) return "origin/main";
    if (args.includes("--show-current")) return "main";
    if (args.includes("--porcelain")) return "";
    throw new Error(`Unexpected git args: ${args.join(" ")}`);
  };

  resolveGitState(spyRunner);

  assert.equal(capturedBudgets.length, 3, "exactly 3 probes must be called");
  for (const budget of capturedBudgets) {
    assert.ok(
      typeof budget === "number" && budget > 0,
      `each probe must receive a positive numeric timeoutMs, got ${JSON.stringify(budget)}`
    );
  }
  // First probe must not exceed the 5000ms budget (+ 1ms tolerance for clock)
  assert.ok(capturedBudgets[0] <= 5001, `first probe timeout must not exceed TIMEOUT_MS: ${capturedBudgets[0]}`);
  // Last probe must not receive more time than first probe (shared deadline shrinks)
  assert.ok(
    capturedBudgets[2] <= capturedBudgets[0],
    `last probe must not get more time than first probe: [${capturedBudgets.join(", ")}]`
  );
});

// ---------------------------------------------------------------------------
// Finding 3 remediation (RED) — hostile branch name sanitized in composeAdvisory
// ---------------------------------------------------------------------------

test("composeAdvisory: hostile branch name — control chars stripped, whitespace collapsed, truncated at 120 chars", () => {
  // Branch name with control characters including ESC (ANSI injection attempt)
  const hostile = "main\x00\x1f\x1b[31mred\x1b[0m\r\ncontinued";
  const result = composeAdvisory(true, false, hostile);

  // Control characters must be stripped
  assert.ok(!/[\x00-\x1f\x7f]/.test(result), "control chars must be removed from advisory");
  // The advisory must still fire (it is not empty)
  assert.ok(result.length > 0, "advisory must be non-empty even for hostile branch name");
});

test("composeAdvisory: long branch name is truncated to at most 120 chars + ellipsis", () => {
  const longBranch = "a".repeat(200);
  const result = composeAdvisory(true, false, longBranch);
  // The full 200-char branch name must NOT appear verbatim in the output
  assert.ok(!result.includes(longBranch), "200-char branch name must be truncated");
  // The advisory must still mention the truncated name
  assert.ok(result.includes("a".repeat(120)) || result.includes("…"), "truncated advisory must contain 120 'a' chars or ellipsis");
});
